File: Binary_Search_Tree/java/search_node.java
// Binary Search Tree 
// JAVA

//  Visual Representation of the binary search tree
//                 10
//            /         \ 
//          5              17
//       /    \           /   \ 
//      2      7        14     22
//     / \    / \      / \     / \       
//    1   4  6   9   11   16 22   27  

class BinarySearchTree_search {
    class Node
    {
        int key;
        Node left, right;
        public Node(int item)
        {
            key = item;
            left = right = null;
        }
    }
    Node root;
    BinarySearchTree_search()
    {
        root = null;
    }

    void insert(int key)
    {
        root = insertRec(root, key);
    }

    Node insertRec(Node root, int key)
    {

        if (root == null)
        {
            root = new Node(key);
            return root;
        }

        if (key < root.key)
            root.left = insertRec(root.left, key);
        else if (key > root.key)
            root.right = insertRec(root.right, key);

        return root;
    }
    void searchKey(int key) {
        root = search(root, key);
    }

    Node search(Node root, int key)
    {
        if (root==null || root.key==key)
        {
            if(root == null)
                {System.out.println("\nNot Found"); 
                System.exit(0);}
            else if(root.key != key)
                {System.out.println("\nNot Found");
                System.exit(0);}
            else 
                {System.out.println("\nFound");
                System.exit(0);}
        }
        if (root.key < key)
            return search(root.right, key);

        return search(root.left, key);
    }

    void inorder()
    {
        inorderRec(root);
    }

    void inorderRec(Node root)
    {
        if (root != null) {
            inorderRec(root.left);
            System.out.println(root.key);
            inorderRec(root.right);
        }
    }


    public static void main(String[] args)
    {
        BinarySearchTree_search tree = new BinarySearchTree_search();
        

        tree.insert(10);

        tree.insert(2);
        tree.insert(1);
        tree.insert(4);
        tree.insert(9);

        tree.insert(7);
        tree.insert(6);
        tree.insert(5);
        tree.insert(9);
        tree.insert(11);
        tree.insert(14);
        tree.insert(19);
        tree.insert(17);

        tree.insert(16);
        tree.insert(27);
        tree.insert(22);


        // tree.inorder();

        

        tree.searchKey(5);


        
        }
}

File: ACADEMIC CODES/Binary_Search_Tree/javascript/search_node.js
// JavaScript implementation
// Search Operation


// Visual Representation of the binary search tree
//                10
//           /          \
//         5              17
//      /    \           /   \
//     2      7        14     22
//    / \    / \      / \     / \
//   1   4  6   9   11   16 22   27

// Node class
class Node
{
    constructor(data)
    {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}



// Binary Search tree class
class BinarySearchTree
{
    constructor()
    {
        this.root = null;
    }

    // Insert Operation
    insert(data)
    {
        var newNode = new Node(data);
        if(this.root === null)
            this.root = newNode;
        else
            this.insertNode(this.root, newNode);
    }

    insertNode(node, newNode)
    {
        if(newNode.data < node.data)
        {
            if(node.left === null)
                node.left = newNode;
            else
                this.insertNode(node.left, newNode); 
        }

        else
        {
            if(node.right === null)
                node.right = newNode;
            else
                this.insertNode(node.right,newNode);
        }
    }

    // To print in inoder method
    inorder(node)
    {
        if(node !== null)
        {
            this.inorder(node.left);
            console.log(node.data);
            this.inorder(node.right);
        }
    }

    getRootNode()
    {
        return this.root;
    }

    // Searching Element
    search(node, data)
    {
        if(node === null)
            return null;

        else if(data < node.data)
            return this.search(node.left, data);

        else if(data > node.data)
            return this.search(node.right, data);

        else
            return node;
    }

}







var BST = new BinarySearchTree();

// Inserting Elements
BST.insert(10);

BST.insert(2);
BST.insert(1);

BST.insert(4);
BST.insert(7);

BST.insert(9);
BST.insert(6);

BST.insert(5);
BST.insert(11);

BST.insert(14);
BST.insert(19);

BST.insert(17);
BST.insert(16);

BST.insert(27);
BST.insert(22);



var root = BST.getRootNode();
// BST.inorder(root);

var a = BST.search(root,5)


if(a === null)
    console.log('\nNot Found');
else if(a.data !== 5)
    console.log('\nNot Found');
else 
    console.log('\nFound');

